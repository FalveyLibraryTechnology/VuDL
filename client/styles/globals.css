:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
