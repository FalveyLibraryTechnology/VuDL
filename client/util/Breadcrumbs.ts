export interface TreeNode {
    pid: string;
    title: string;
    parents: Array<TreeNode>;
}

export interface TreeData {
    topNodes: Array<string>;
    records: Record<string, TreeNode>;
    childLookups: Record<string, Set<string>>;
}

interface BreadcrumbTrail {
    pid: string;
    path: Array<TreeNode>;
}

/**
 * Analyze the raw breadcrumb data, identifying top-level nodes (i.e. nodes with
 * no parents) and creating lookup tables for children. This makes it easier to
 * render breadcrumbs from left to right and to find all relevant trails.
 */
export function processBreadcrumbData(data: TreeNode): TreeData {
    const queue: Array<TreeNode> = [data];
    const topNodes: Set<string> = new Set(); // use set to avoid duplicates
    const childLookups: Record<string, Set<string>> = {};
    const records: Record<string, TreeNode> = {};
    while (queue.length > 0) {
        const current = queue.shift();
        if (current.parents.length === 0) {
            topNodes.add(current.pid);
        }
        records[current.pid] = { pid: current.pid, title: current.title, parents: [] };
        current.parents.forEach((parent) => {
            queue.push(parent);
            if (typeof childLookups[parent.pid] === "undefined") {
                childLookups[parent.pid] = new Set([current.pid]);
            } else {
                childLookups[parent.pid].add(current.pid);
            }
        });
    }
    return {
        topNodes: Array.from(topNodes),
        records,
        childLookups,
    };
}

export function generateBreadcrumbTrails(treeData: TreeData, pid: string) {
    // BFS from top (root id) to target pid
    const queue: Array<BreadcrumbTrail> = (treeData.topNodes ?? []).map((rootId) => {
        return {
            pid: rootId,
            path: [],
        };
    });
    const result: Array<Array<TreeNode>> = [];
    while (queue.length > 0) {
        const current = queue.shift();
        const record = treeData.records[current.pid] ?? {};
        const path = current.path;
        path.push(record);
        (treeData.childLookups[current.pid] ?? []).forEach((childPid: string) => {
            // At target
            if (childPid === pid) {
                result.push(path);
            } else {
                // Add to queue for more
                queue.push({
                    pid: childPid,
                    path: [...path], // clone array to avoid multiple references to the same array
                });
            }
        });
    }
    // Even if no trails were found, at least return a single empty array so that we can
    // display the "Edit Home" link when we process the data.
    return result.length == 0 ? [[]] : result;
}

