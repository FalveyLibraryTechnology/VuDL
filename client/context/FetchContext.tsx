import React, { createContext, useContext, useReducer } from "react";
import { authApiUrl, ingestApiUrl, loginUrl } from "../util/routes";
import PropTypes from "prop-types";

/**
 * Pass a shared entity to react components,
 * specifically a way to make api requests.
 */
const fetchContextParams = {
    token: typeof window !== 'undefined' ? sessionStorage.getItem("token") : null,
};
const FetchContext = createContext({});

/**
 * Update the shared states of react components.
 */
const fetchReducer = (state, { type, payload }) => {
    switch (type) {
        case "UPDATE_TOKEN":
            sessionStorage.setItem("token", payload);
            return {
                ...state,
                token: payload,
            };
        default:
            console.error(`fetch action type: ${type} does not exist`);
            return state;
    }
};

export const FetchContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(fetchReducer, fetchContextParams);
    const value = { state, dispatch };
    return <FetchContext.Provider value={value}>{children}</FetchContext.Provider>;
};

export const useFetchContext = () => {
    const {
        state: { token },
        dispatch,
    } = useContext(FetchContext);

    /**
     * Get default fetch parameters. If params or headers exist,
     * overload the associated keys.
     */
    const fetchParams = (params = {}, headers = {}) => ({
        credentials: "include",
        method: "GET",
        mode: "cors",
        ...params,
        headers,
    });

    /**
     * Query a fresh token to authorize user to make api requests.
     */
    const refreshToken = async () => {
        try {
            const response = await fetch(`${authApiUrl}/token/mint`, fetchParams());
            if (response.ok) {
                const token = await response.json();
                dispatch({
                    type: "UPDATE_TOKEN",
                    payload: token,
                });
                return token;
            }
        } catch (error) {
            console.error(error);
        }
        window.location.href = `${loginUrl}?referer=${encodeURIComponent(window.location.href)}`;
        return false;
    };

    /**
     * Return a customizable response when making a request.
     * If first response return a 401 unauthorized status,
     * attempt to refresh token and call the request again.
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const makeRequest = async (url, params = {}, headers = {}) => {
        const response = await fetch(
            url,
            fetchParams(params, {
                ...headers,
                Authorization: `Token ${token}`,
            })
        );
        if (response.status == 401) {
            const token = await refreshToken();
            return token
                ? await fetch(
                      url,
                      fetchParams(params, {
                          ...headers,
                          Authorization: `Token ${token}`,
                      })
                  )
                : false;
        }
        return response;
    };

    /**
     * Return a json response when making a request
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const fetchJSON = async (url = ingestApiUrl, params = {}, headers = {}) => {
        const response = await makeRequest(url, params, headers);
        if (response?.ok) {
            return await response.json();
        }
        throw new Error(response?.statusText ?? "Could not make request");
    };

    /**
     * Return a text response when making a request
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const fetchText = async (url = ingestApiUrl, params = {}, headers = {}) => {
        const response = await makeRequest(url, params, headers);
        if (response.ok) {
            return await response.text();
        }
        throw new Error(response.statusText);
    };

    return {
        state: {
            token,
        },
        action: {
            fetchJSON,
            fetchText,
            makeRequest,
        },
    };
};

FetchContextProvider.propTypes = {
    children: PropTypes.node,
};

export default { FetchContextProvider, useFetchContext };
