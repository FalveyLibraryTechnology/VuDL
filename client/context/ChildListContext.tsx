import React, { createContext, useContext, useReducer } from "react";
import { getObjectChildrenUrl } from "../util/routes";
import { useFetchContext } from "./FetchContext";

interface Children {
    numFound?: number;
    start?: number;
    docs?: Record<string, string>[];
}

interface ChildListState {
    childStorage: Record<string, Children>;
}

const childListContextParams: ChildListState = {
    childStorage: {},
};

const ChildListContext = createContext({});

const reducerMapping: Record<string, string> = {
};

/**
 * Update the shared states of react components.
 */
const childListReducer = (state: ChildListState, { type, payload }: { type: string, payload: unknown}) => {
    if (type === "ADD_TO_CHILD_STORAGE") {
        const { key, children } = payload;
        const childStorage = {
            ...state.childStorage,
        };
        childStorage[key] = children;
        return {
            ...state,
            childStorage
        };
    } else if (Object.keys(reducerMapping).includes(type)){
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    } else {
        console.error(`fetch action type: ${type} does not exist`);
        return state;
    }
};

export const ChildListContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(childListReducer, childListContextParams);
    const value = { state, dispatch };
    return <ChildListContext.Provider value={value}>{children}</ChildListContext.Provider>;
};

export const useChildListContext = () => {
    const {
        action: {
            fetchJSON
        }
    }= useFetchContext();
    const {
        state: {
            childStorage
        },
        dispatch,
    } = useContext(ChildListContext);

    const addToChildStorage = (key: string, children: Children) => {
        dispatch({
            type: "ADD_TO_CHILD_STORAGE",
            payload: { key, children },
        });
    };

    const getChildStorageKey = (pid: string, page: number, pageSize: number): string => {
        return `${pid}_${page}_${pageSize}`;
    }

    const loadChildrenIntoStorage = async (pid: string, page: number, pageSize: number) => {
        const key = getChildStorageKey(pid, page, pageSize);
        const url = getObjectChildrenUrl(pid, (page - 1) * pageSize, pageSize);
        try {
            addToChildStorage(key, await fetchJSON(url));
        } catch (e) {
            console.error("Problem fetching tree data from " + url);
        }
    }

    return {
        state: {
            childStorage,
        },
        action: {
            getChildStorageKey,
            loadChildrenIntoStorage,
        },
    };
}

export default {
    ChildListContextProvider,
    useChildListContext
}