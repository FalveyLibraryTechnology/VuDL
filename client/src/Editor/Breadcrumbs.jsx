import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useFetchContext } from "../FetchContext";
import { apiUrl } from "../routes";
import { Link } from "react-router-dom";

const Breadcrumbs = ({ pid = null }) => {
    const {
        action: { fetchJSON },
    } = useFetchContext();
    const [breadcrumbs, setBreadcrumbs] = useState([]);

    useEffect(() => {
        async function loadData() {
            let data = [];
            const url = apiUrl + "/edit/breadcrumbs" + (pid === null ? "" : "/" + pid);
            try {
                data = await fetchJSON(url);
            } catch (e) {
                console.error("Problem fetching breadcrumb data from " + url);
            }
            setBreadcrumbs(data);
        }
        loadData();
    }, []);
    const contents = (breadcrumbs.parents ?? []).map((breadcrumb) => {
        return (
            <li key={(breadcrumb.pid ?? "root") + breadcrumb.pid}>
                <Link to={"/edit/object/" + breadcrumb.pid}>{breadcrumb.title}</Link>
            </li>
        );
    });
    return <ul>{contents}</ul>;
};

Breadcrumbs.propTypes = {
    pid: PropTypes.string,
};

export default Breadcrumbs;
