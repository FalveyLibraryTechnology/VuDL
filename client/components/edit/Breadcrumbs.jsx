import styles from "./Breadcrumbs.module.css";
import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useFetchContext } from "../../context/FetchContext";
import { apiUrl } from "../../util/routes";
import Link from "next/link";

/* TODO: use this interface when we convert to Typescript:
interface TreeData {
    topNodes: Array<string>,
    records: Record<string, Record<string, string>>,
    childLookups: Record<string, Array<string>>,
}
 */

const Breadcrumbs = ({ pid = null }) => {
    const {
        action: { fetchJSON },
    } = useFetchContext();
    const [treeData, setTreeData] = useState([]);

    /**
     * Analyze the raw breadcrumb data, identifying top-level nodes (i.e. nodes with
     * no parents) and creating lookup tables for children. This makes it easier to
     * render breadcrumbs from left to right and to find all relevant trails.
     */
    function processBreadcrumbData(data) {
        const queue = [data];
        const topNodes = [];
        const childLookups = {};
        const records = {};
        while (queue.length > 0) {
            const current = queue.shift();
            if (current.parents.length === 0) {
                topNodes.push(current.pid);
            }
            records[current.pid] = { pid: current.pid, title: current.title };
            current.parents.forEach((parent) => {
                queue.push(parent);
                if (typeof childLookups[parent.pid] === "undefined") {
                    childLookups[parent.pid] = [current.pid];
                } else {
                    childLookups[parent.pid].push(current.pid);
                }
            });
        }
        return {
            topNodes: [...new Set(topNodes)], // deduplicate
            records,
            childLookups,
        };
    }

    useEffect(() => {
        async function loadData() {
            let data = [];
            const url = apiUrl + "/edit/object/parents/" + encodeURIComponent(pid);
            try {
                data = processBreadcrumbData(await fetchJSON(url));
            } catch (e) {
                console.error("Problem fetching breadcrumb data from " + url);
            }
            setTreeData(data);
        }
        loadData();
    }, []);

    function generateBreadcrumbTrails(treeData, pid) {
        // BFS from top (root id) to target pid
        const queue = [];
        (treeData.topNodes ?? []).forEach((rootId) => {
            queue.push({
                pid: rootId,
                path: [],
            });
        });
        const result = [];
        while (queue.length > 0) {
            const current = queue.shift();
            const record = treeData.records[current.pid] ?? {};
            const path = current.path;
            path.push(record);
            (treeData.childLookups[current.pid] ?? []).forEach((childPid) => {
                // At target
                if (childPid === pid) {
                    result.push(path);
                } else {
                    // Add to queue for more
                    queue.push({
                        pid: childPid,
                        path: [...path], // clone array to avoid multiple references to the same array
                    });
                }
            });
        }
        // Even if no trails were found, at least return a single empty array so that we can
        // display the "Edit Home" link when we process the data.
        return result.length == 0 ? [[]] : result;
    }

    const allTrails = generateBreadcrumbTrails(treeData, pid);
    const contents = allTrails.map((trail, trailIndex) => {
        const breadcrumbs = trail.map((breadcrumb) => {
            return (
                <li key={"breacrumb_" + breadcrumb.pid + "_" + trailIndex}>
                    <Link href={"/edit/object/" + breadcrumb.pid}>{breadcrumb.title}</Link>
                </li>
            );
        });
        breadcrumbs.unshift(
            <li key={"breadcrumb_home_" + trailIndex}>
                <Link href="/edit">Edit Home</Link>
            </li>
        );
        return <ul className={styles.breadcrumb} key={"breadcrumbs" + "_" + trailIndex}>{breadcrumbs}</ul>;
    });
    return <>{contents}</>;
};

Breadcrumbs.propTypes = {
    pid: PropTypes.string,
};

export default Breadcrumbs;
