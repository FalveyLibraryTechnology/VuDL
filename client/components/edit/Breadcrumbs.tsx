import styles from "./Breadcrumbs.module.css";
import React, { useEffect } from "react";
import { TreeNode, processBreadcrumbData } from "../../util/Breadcrumbs";
import { useEditorContext } from "../../context/EditorContext";
import Link from "next/link";

interface BreadcrumbsProps {
    pid?: string | null;
}

const Breadcrumbs = ({ pid = null }: BreadcrumbsProps): React.ReactElement => {
    const {
        state: { parentDetailsStorage },
        action: { loadParentDetailsIntoStorage },
    } = useEditorContext();
    const loaded = Object.prototype.hasOwnProperty.call(parentDetailsStorage, pid);

    useEffect(() => {
        if (!loaded && pid !== null) {
            loadParentDetailsIntoStorage(pid);
        }
    }, [loaded]);

    // Special case: no PID, we're at the top level:
    if (pid === null) {
        return (
            <ul className={styles.breadcrumb}>
                <li>
                    <Link href="/">Main Menu</Link>
                </li>
            </ul>
        );
    }

    const treeData: Array<Array<TreeNode>> =
        loaded && pid ? processBreadcrumbData(parentDetailsStorage[pid]).paths : [];

    const contents = treeData.map((trail, trailIndex: number) => {
        const breadcrumbs = trail.map((breadcrumb) => {
            return (
                <li key={"breadcrumb_" + breadcrumb.pid + "_" + trailIndex}>
                    <Link href={"/edit/object/" + breadcrumb.pid}>{breadcrumb.title}</Link>
                </li>
            );
        });
        breadcrumbs.unshift(
            <li key={"breadcrumb_home_" + trailIndex}>
                <Link href="/edit">Edit Home</Link>
            </li>
        );
        breadcrumbs.unshift(
            <li key={"breadcrumb_mainmenu_" + trailIndex}>
                <Link href="/">Main Menu</Link>
            </li>
        );
        return (
            <ul className={styles.breadcrumb} key={"breadcrumbs_" + trailIndex}>
                {breadcrumbs}
            </ul>
        );
    });
    return <>{contents}</>;
};

export default Breadcrumbs;
