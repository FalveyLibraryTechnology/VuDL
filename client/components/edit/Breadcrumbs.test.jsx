import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { waitFor } from "@testing-library/react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import Breadcrumbs from "./Breadcrumbs";
import { FetchContextProvider } from "../../context/FetchContext";

describe("Breadcrumb", () => {
    let props;
    let lastRequestUrl;
    let breadcrumbResponse = {};

    beforeEach(() => {
        props = { pid: "foo:1234" };
        global.fetch = jest.fn((url) => {
            lastRequestUrl = url;
            return {
                ok: true,
                status: 200,
                json: async function () {
                    return breadcrumbResponse;
                },
            };
        });
    });

    async function runStandardSnapshotTest() {
        const wrapper = mount(
            <FetchContextProvider>
                <Breadcrumbs {...props} />
            </FetchContextProvider>
        );
        await waitFor(() => expect(global.fetch).toHaveBeenCalledTimes(1));
        expect(lastRequestUrl).toEqual("http://localhost:9000/api/edit/object/parents/foo%3A1234");
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    }

    function getObject(pid, title, parents = []) {
        return { pid, title, parents };
    }

    it("renders using ajax-loaded object data (no parents)", async () => {
        breadcrumbResponse = getObject("foo:1234", "Fake Title");
        await runStandardSnapshotTest();
    });

    it("renders using ajax-loaded object data (simple case)", async () => {
        const parent = getObject("foo:1233", "Fake Parent");
        breadcrumbResponse = getObject("foo:1234", "Fake Title", [parent]);
        await runStandardSnapshotTest();
    });

    it("renders using ajax-loaded object data (many parents)", async () => {
        const parents = [];
        for (let $x = 0; $x < 15; $x++) {
            parents.push(getObject("foo:" + $x, "Fake Parent " + $x));
        }
        breadcrumbResponse = getObject("foo:1234", "Fake Title", parents);
        await runStandardSnapshotTest();
    });

    it("renders using ajax-loaded object data (deep parent chain)", async () => {
        let parents = [];
        for (let $x = 0; $x < 15; $x++) {
            parents = [getObject("foo:" + $x, "Fake Parent " + $x, parents)];
        }
        breadcrumbResponse = getObject("foo:1234", "Fake Title", parents);
        await runStandardSnapshotTest();
    });

    it("renders using ajax-loaded object data (multiple parents with common grandparent)", async () => {
        const grandparent = getObject("foo:1232", "Fake Grandparent");
        const parent1 = getObject("foo:1233", "Fake Parent 1", [grandparent]);
        const parent2 = getObject("foo:1231", "Fake Parent 2", [grandparent]);
        breadcrumbResponse = getObject("foo:1234", "Fake Title", [parent1, parent2]);
        await runStandardSnapshotTest();
    });
});
