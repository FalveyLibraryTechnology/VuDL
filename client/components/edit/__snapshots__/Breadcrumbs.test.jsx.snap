// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb renders using ajax-loaded object data (deep parent chain) 1`] = `
<FetchContextProvider>
  <Breadcrumbs
    pid="foo:1234"
  >
    <ul
      key="breadcrumbs_0"
    >
      <li
        key="foo:0_0"
      >
        <Link
          href="/edit/object/foo:0"
        >
          <a
            href="/edit/object/foo:0"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 0
          </a>
        </Link>
      </li>
      <li
        key="foo:1_0"
      >
        <Link
          href="/edit/object/foo:1"
        >
          <a
            href="/edit/object/foo:1"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 1
          </a>
        </Link>
      </li>
      <li
        key="foo:2_0"
      >
        <Link
          href="/edit/object/foo:2"
        >
          <a
            href="/edit/object/foo:2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 2
          </a>
        </Link>
      </li>
      <li
        key="foo:3_0"
      >
        <Link
          href="/edit/object/foo:3"
        >
          <a
            href="/edit/object/foo:3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 3
          </a>
        </Link>
      </li>
      <li
        key="foo:4_0"
      >
        <Link
          href="/edit/object/foo:4"
        >
          <a
            href="/edit/object/foo:4"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 4
          </a>
        </Link>
      </li>
      <li
        key="foo:5_0"
      >
        <Link
          href="/edit/object/foo:5"
        >
          <a
            href="/edit/object/foo:5"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 5
          </a>
        </Link>
      </li>
      <li
        key="foo:6_0"
      >
        <Link
          href="/edit/object/foo:6"
        >
          <a
            href="/edit/object/foo:6"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 6
          </a>
        </Link>
      </li>
      <li
        key="foo:7_0"
      >
        <Link
          href="/edit/object/foo:7"
        >
          <a
            href="/edit/object/foo:7"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 7
          </a>
        </Link>
      </li>
      <li
        key="foo:8_0"
      >
        <Link
          href="/edit/object/foo:8"
        >
          <a
            href="/edit/object/foo:8"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 8
          </a>
        </Link>
      </li>
      <li
        key="foo:9_0"
      >
        <Link
          href="/edit/object/foo:9"
        >
          <a
            href="/edit/object/foo:9"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 9
          </a>
        </Link>
      </li>
      <li
        key="foo:10_0"
      >
        <Link
          href="/edit/object/foo:10"
        >
          <a
            href="/edit/object/foo:10"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 10
          </a>
        </Link>
      </li>
      <li
        key="foo:11_0"
      >
        <Link
          href="/edit/object/foo:11"
        >
          <a
            href="/edit/object/foo:11"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 11
          </a>
        </Link>
      </li>
      <li
        key="foo:12_0"
      >
        <Link
          href="/edit/object/foo:12"
        >
          <a
            href="/edit/object/foo:12"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 12
          </a>
        </Link>
      </li>
      <li
        key="foo:13_0"
      >
        <Link
          href="/edit/object/foo:13"
        >
          <a
            href="/edit/object/foo:13"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 13
          </a>
        </Link>
      </li>
      <li
        key="foo:14_0"
      >
        <Link
          href="/edit/object/foo:14"
        >
          <a
            href="/edit/object/foo:14"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 14
          </a>
        </Link>
      </li>
    </ul>
  </Breadcrumbs>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (many parents) 1`] = `
<FetchContextProvider>
  <Breadcrumbs
    pid="foo:1234"
  >
    <ul
      key="breadcrumbs_0"
    >
      <li
        key="foo:0_0"
      >
        <Link
          href="/edit/object/foo:0"
        >
          <a
            href="/edit/object/foo:0"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 0
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_1"
    >
      <li
        key="foo:1_1"
      >
        <Link
          href="/edit/object/foo:1"
        >
          <a
            href="/edit/object/foo:1"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 1
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_2"
    >
      <li
        key="foo:2_2"
      >
        <Link
          href="/edit/object/foo:2"
        >
          <a
            href="/edit/object/foo:2"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 2
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_3"
    >
      <li
        key="foo:3_3"
      >
        <Link
          href="/edit/object/foo:3"
        >
          <a
            href="/edit/object/foo:3"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 3
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_4"
    >
      <li
        key="foo:4_4"
      >
        <Link
          href="/edit/object/foo:4"
        >
          <a
            href="/edit/object/foo:4"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 4
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_5"
    >
      <li
        key="foo:5_5"
      >
        <Link
          href="/edit/object/foo:5"
        >
          <a
            href="/edit/object/foo:5"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 5
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_6"
    >
      <li
        key="foo:6_6"
      >
        <Link
          href="/edit/object/foo:6"
        >
          <a
            href="/edit/object/foo:6"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 6
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_7"
    >
      <li
        key="foo:7_7"
      >
        <Link
          href="/edit/object/foo:7"
        >
          <a
            href="/edit/object/foo:7"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 7
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_8"
    >
      <li
        key="foo:8_8"
      >
        <Link
          href="/edit/object/foo:8"
        >
          <a
            href="/edit/object/foo:8"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 8
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_9"
    >
      <li
        key="foo:9_9"
      >
        <Link
          href="/edit/object/foo:9"
        >
          <a
            href="/edit/object/foo:9"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 9
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_10"
    >
      <li
        key="foo:10_10"
      >
        <Link
          href="/edit/object/foo:10"
        >
          <a
            href="/edit/object/foo:10"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 10
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_11"
    >
      <li
        key="foo:11_11"
      >
        <Link
          href="/edit/object/foo:11"
        >
          <a
            href="/edit/object/foo:11"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 11
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_12"
    >
      <li
        key="foo:12_12"
      >
        <Link
          href="/edit/object/foo:12"
        >
          <a
            href="/edit/object/foo:12"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 12
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_13"
    >
      <li
        key="foo:13_13"
      >
        <Link
          href="/edit/object/foo:13"
        >
          <a
            href="/edit/object/foo:13"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 13
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_14"
    >
      <li
        key="foo:14_14"
      >
        <Link
          href="/edit/object/foo:14"
        >
          <a
            href="/edit/object/foo:14"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 14
          </a>
        </Link>
      </li>
    </ul>
  </Breadcrumbs>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (multiple parents with common grandparent) 1`] = `
<FetchContextProvider>
  <Breadcrumbs
    pid="foo:1234"
  >
    <ul
      key="breadcrumbs_0"
    >
      <li
        key="foo:1232_0"
      >
        <Link
          href="/edit/object/foo:1232"
        >
          <a
            href="/edit/object/foo:1232"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Grandparent
          </a>
        </Link>
      </li>
      <li
        key="foo:1233_0"
      >
        <Link
          href="/edit/object/foo:1233"
        >
          <a
            href="/edit/object/foo:1233"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 1
          </a>
        </Link>
      </li>
    </ul>
    <ul
      key="breadcrumbs_1"
    >
      <li
        key="foo:1232_1"
      >
        <Link
          href="/edit/object/foo:1232"
        >
          <a
            href="/edit/object/foo:1232"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Grandparent
          </a>
        </Link>
      </li>
      <li
        key="foo:1231_1"
      >
        <Link
          href="/edit/object/foo:1231"
        >
          <a
            href="/edit/object/foo:1231"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent 2
          </a>
        </Link>
      </li>
    </ul>
  </Breadcrumbs>
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (no parents) 1`] = `
<FetchContextProvider>
  <Breadcrumbs
    pid="foo:1234"
  />
</FetchContextProvider>
`;

exports[`Breadcrumb renders using ajax-loaded object data (simple case) 1`] = `
<FetchContextProvider>
  <Breadcrumbs
    pid="foo:1234"
  >
    <ul
      key="breadcrumbs_0"
    >
      <li
        key="foo:1233_0"
      >
        <Link
          href="/edit/object/foo:1233"
        >
          <a
            href="/edit/object/foo:1233"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Fake Parent
          </a>
        </Link>
      </li>
    </ul>
  </Breadcrumbs>
</FetchContextProvider>
`;
